django/db/models/base.py

```diff
@@ -362,9 +362,18 @@ def _prepare(cls):
         # Set the name of _meta.indexes. This can't be done in
         # Options.contribute_to_class() because fields haven't been added to
         # the model at that point.
+        #
+        # if the name is set with dict argument,
+        # name will be set by set_name_with_dict for indexes and constraints.
         for index in cls._meta.indexes:
             if not index.name:
                 index.set_name_with_model(cls)
+            else:
+                index.set_name_with_dict(cls)
```

Comment:
`set_name_with_dict`, this method name confuses me, I feel like I need to provide a dictionary as argument. Maybe something like `format_name_with_model` can be better?

Reply:
generate a name was tough for me :) I changed the name as your comment. Maybe the third opinion would be nice.